shared-module/displayio/FourWire.c

```diff
@@ -113,19 +117,52 @@ bool common_hal_displayio_fourwire_begin_transaction(mp_obj_t obj) {
 
 void common_hal_displayio_fourwire_send(mp_obj_t obj, display_byte_type_t data_type,
     display_chip_select_behavior_t chip_select, const uint8_t *data, uint32_t data_length) {
+    uint8_t buffer = 0;
     displayio_fourwire_obj_t *self = MP_OBJ_TO_PTR(obj);
-    common_hal_digitalio_digitalinout_set_value(&self->command, data_type == DISPLAY_DATA);
-    if (chip_select == CHIP_SELECT_TOGGLE_EVERY_BYTE) {
-        // Toggle chip select after each command byte in case the display driver
-        // IC latches commands based on it.
+    if (self->command.base.type == &mp_type_NoneType) {
+        uint8_t bits = 0;
+        uint8_t dc = (data_type == DISPLAY_DATA);
+
         for (size_t i = 0; i < data_length; i++) {
-            common_hal_busio_spi_write(self->bus, &data[i], 1);
+            bits = (bits + 1) % 8;
+
+            if (bits == 0) {
+                // send the previous byte and the dc bit
+                buffer = (buffer << 1) | dc;
+                common_hal_busio_spi_write(self->bus, &buffer, 1);
+                // send the current byte
+                common_hal_busio_spi_write(self->bus, &data[i], 1);
+            } else {
+                // send remaining bits from previous byte, dc and beginning of current byte
+                buffer = (buffer << (9 - bits)) | (dc << (8 - bits)) | (data[i] >> bits);
+                common_hal_busio_spi_write(self->bus, &buffer, 1);
+            }
+            // save the current byte
+            buffer = data[i];
+        }
+        if (bits > 0) {
+            // send any remaining bits
+            buffer = buffer << (8 - bits);
+            common_hal_busio_spi_write(self->bus, &buffer, 1);
+            // toggle CS to discard superfluous bits
```

Comment:
Is this timing sensitive? I'm a little worried it'll depend on the SoC's clock speeds.

Please document this trick at the start of this code.

Reply:
It's just a normal CS wiggle, the same as what happens when you set `CHIP_SELECT_TOGGLE_EVERY_BYTE`, except it only ever happens at the end of a block of data, not at every byte.

I noticed during debugging that with very slow SPI clock speeds (around 1000), the CS is toggled before last bit of the DMA transfer finishes, but I suspect that is a general bug in the SAMD implementation of SPI — the write call waits for the last bit to begin, not for it to end. With normal speeds that are more practical there is no such quirk. I tested the display at 40MHz without problems.

I will add a more elaborate explanation.